components/tikv_util/src/config.rs

```diff
@@ -457,17 +457,104 @@ impl<'de> Deserialize<'de> for ReadableDuration {
     }
 }
 
+fn canonicalize_fallback<P: AsRef<Path>>(path: P) -> std::io::Result<PathBuf> {
+    fn normalize(path: &Path) -> PathBuf {
+        use std::path::Component;
+        let mut components = path.components().peekable();
+        let mut ret = PathBuf::new();
+
+        while let Some(c @ (Component::Prefix(..) | Component::RootDir)) =
+            components.peek().cloned()
+        {
+            components.next();
+            ret.push(c.as_os_str());
+        }
+
+        let mut normal_nodes = 0;
+        for component in components {
+            match component {
+                Component::Prefix(..) | Component::RootDir => unreachable!(),
+                Component::CurDir => {}
+                c @ Component::ParentDir => {
+                    if normal_nodes > 0 {
+                        ret.pop();
+                    } else {
+                        ret.push(c.as_os_str());
+                    }
+                }
+                Component::Normal(c) => {
+                    ret.push(c);
+                    normal_nodes += 1;
+                }
+            }
+        }
+        ret
+    }
+    fn try_canonicalize_normalized_path(path: &Path) -> std::io::Result<PathBuf> {
+        use std::path::Component;
+        let mut components = path.components().peekable();
+        let mut should_canonicalize = true;
+        let mut ret = if path.is_relative() {
+            Path::new(".").canonicalize()?
+        } else {
+            PathBuf::new()
+        };
+
+        while let Some(c @ (Component::Prefix(..) | Component::RootDir)) =
+            components.peek().cloned()
+        {
+            components.next();
+            ret.push(c.as_os_str());
+        }
+        // Normalize will only preserve leading ParentDir.
+        while let Some(Component::ParentDir) = components.peek().cloned() {
+            components.next();
+            ret.pop();
+        }
+
+        for component in components {
+            match component {
+                Component::Normal(c) => {
+                    ret.push(c);
+                    // We try to canonicalize a longest path based on fs info.
+                    if should_canonicalize {
+                        match ret.as_path().canonicalize() {
+                            Ok(path) => {
+                                ret = path;
+                            }
+                            Err(e) if e.kind() == std::io::ErrorKind::NotFound => {
+                                should_canonicalize = false;
+                            }
+                            other => return other,
+                        }
+                    }
+                }
+                Component::Prefix(..)
+                | Component::RootDir
+                | Component::ParentDir
+                | Component::CurDir => unreachable!(),
```

Comment:
`ParentDir` is no longer unreachable when `nomalize` needs to reserve the prefix `ParentDir`.
```suggestion
                Component::ParentDir => {
                    ret.pop();
                },
                Component::Prefix(..)
                | Component::RootDir
                | Component::CurDir => unreachable!(),
```